sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala

```diff
@@ -276,15 +276,15 @@ object BooleanSimplification extends Rule[LogicalPlan] with PredicateHelper {
       case a And b if a.semanticEquals(b) => a
       case a Or b if a.semanticEquals(b) => a
 
-      case a And (b Or c) if Not(a).semanticEquals(b) => And(a, c)
-      case a And (b Or c) if Not(a).semanticEquals(c) => And(a, b)
-      case (a Or b) And c if a.semanticEquals(Not(c)) => And(b, c)
-      case (a Or b) And c if b.semanticEquals(Not(c)) => And(a, c)
-
-      case a Or (b And c) if Not(a).semanticEquals(b) => Or(a, c)
-      case a Or (b And c) if Not(a).semanticEquals(c) => Or(a, b)
-      case (a And b) Or c if a.semanticEquals(Not(c)) => Or(b, c)
-      case (a And b) Or c if b.semanticEquals(Not(c)) => Or(a, c)
+      case a And (b Or c) if !a.nullable && Not(a).semanticEquals(b) => And(a, c)
+      case a And (b Or c) if !a.nullable && Not(a).semanticEquals(c) => And(a, b)
+      case (a Or b) And c if !a.nullable && a.semanticEquals(Not(c)) => And(b, c)
+      case (a Or b) And c if !b.nullable && b.semanticEquals(Not(c)) => And(a, c)
+
+      case a Or (b And c) if !a.nullable && Not(a).semanticEquals(b) => Or(a, c)
```

Comment:
these shouldn't be a problem, since if `a` is `true`, then `a Or b` is true, regardless of `b`'s value/nullability, isn't it?

Reply:
the problem is when a is null, c is true

Reply:
Sorry, it is the other case where the change is not needed, right?
`a And (b Or c)` -> `And(a, c)` when `a` is `null`, `And(a, c)` returns `null` (I got a bit confused earlier, sorry).

Reply:
> when a is null, And(a, c) returns null

This is not always the case, `null && false`  is `false`

Reply:
oh, yes you're right, this might be a problem indeed if the expression is inside a `not`. Sorry, thanks.